Data/Pool/Introspection.hs

```diff
@@ -0,0 +1,83 @@
+-- | A variant of "Data.Pool" with introspection capabilities.
+module Data.Pool.Introspection
+  ( -- * Pool
+    Pool
+  , LocalPool
+  , newPool
+
+    -- * Resource management
+  , Resource(..)
+  , AcquisitionMethod(..)
+  , withResource
+  , takeResource
+  , putResource
+  , destroyResource
+  , destroyAllResources
+  ) where
+
+import Control.Concurrent
+import Control.Exception
+import GHC.Clock
+import GHC.Generics (Generic)
+
+import Data.Pool.Internal
+
+-- | A resource taken from the pool along with additional information.
+data Resource a = Resource
+  { resource           :: a
+  , stripeNumber       :: !Int
+  , acquisitionTime    :: !Double
+  , acquisitionMethod  :: !AcquisitionMethod
+  , availableResources :: !Int
+  } deriving (Eq, Show, Generic)
```

Comment:
So this type gives us the resource stats. Notably we get to know how much time was spent on acquiring this resource, whether or not we had to wait on a resource to become available, and how many more resources are currently available in the `LocalPool` stripe.

Comparing with #24, which introduced this type:

```haskell
-- | Stats for a single 'LocalPool'.
data PoolStats = PoolStats {
      highwaterUsage :: Int
    -- ^ Highest usage since last reset.
    , currentUsage   :: Int
    -- ^ Current number of items.
    , takes          :: Int
    -- ^ Number of takes since last reset.
    , creates        :: Int
    -- ^ Number of creates since last reset.
    , createFailures :: Int
    -- ^ Number of creation failures since last reset.
} deriving (Show)
```

We don't get these stats, and that's what we're looking for (at least, in our use at Mercury).

We can get `creates` by modifying the function passed in `newPool` to increment a `TVar` on our side, but that won't be `LocalPool` specific - it'll be limited to the `Pool` as a whole. Probably same with `createFailures`. We cannot get `currentUsage` with the API in `Pool` already, nor can we get `takes`. We are not using `highwaterUsage` since we did not want to reset the pool metrics each time.

Reply:
It's hard to make a set of stats to satisfy everyone and not blow it up :slightly_smiling_face: 

But since everything from the `.Internal` module is exposed, I think writing a version of `takeResource` that would work for you is not a problem?